File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int[][] dp = new int[n+1][n+1];

        for(int i = n-1; i >= 0; --i) {
            for(int j = i-1; j>= -1; --j) {
                int len = dp[i+1][j+1];
                if(j == -1 || nums[i] > nums[j])
                    len = Math.max(len, 1 + dp[i+1][i+1]);
                dp[i][j+1] = len;
            }
        }

        return dp[0][0];
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int f(vector<int> &nums, int idx, int prev, vector<vector<int>> &dp) {
        if(idx >= nums.size()) return 0;
        if(dp[idx][prev+1] != -1) return dp[idx][prev+1];
        
        int p = -1;
        if(prev ==-1 || nums[prev] < nums[idx]) p = 1 + f(nums, idx+1, idx, dp);
        int q = f(nums, idx+1, prev, dp);
        return dp[idx][prev+1] = max(p, q);
    }
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n+1, vector<int>(n+1, -1));
        
        return f(nums, 0, -1, dp);
    }
};